### DISCUSSION TITLE
Very long connection establishment

### DISCUSSION THREAD

**Discussion by Fireshtorm1k** [zeroc-ice/ice#2661] [Q&A] [answered]
I have this slice interface
```
interface Session {
        FluidPackage* createFluidPackage(FluidPackageType fp);
	void release();
    };

interface Application {
		idempotent string getVersion(); 
		idempotent ComponentsDataArray getComponentsDataArray();
		Session* getSession();
	};	
``` 
When i launch client and trying to connect to server, all doing good until i call createFluidPackage. This call reaches the server only after +-45 seconds, and for some reason, he establishes the TCP connection again, despite the fact that it was set up at the beginning.

**Comment by externl** [comment, marked as answer]
Without seeing your code it's hard to say why a new connection is being established. Are you using ACM? 

One potential reason for slow connection establishment could be if your proxy has some unreachable endpoints. Eg. endpoints that are local to the server's internal network but that the client doesn't have access to. Looking at your sever log, your server is listening on public and private network interfaces. 

What are the proxy's endpoints?

**Reply by Fireshtorm1k** [reply]
<details>
  <summary>Client code</summary>
  
```cpp
Ice::ObjectPrx appProxy = communicator->stringToProxy("flowengine:default -h " + ip + " -p " + port + " -t 3000");
ApplicationPrx pApp = pApp.checkedCast(appProxy);
SessionPrx pSession = pApp->getSession();
FluidPackagePrx pFluidPackage = pSession->createFluidPackage(FluidPackageType::Test); // stucks here
```

</details>
<details>
  <summary>Server code</summary>
  
```cpp
  try
    {
        Ice::InitializationData initData;
        initData.properties = Ice::createProperties(argc, argv);
        initData.properties->load("config.server");
        communicator = Ice::initialize(argc, argv, initData);
        BaseObj::adapter = communicator->createObjectAdapter("FlowEngine");
        ApplicationI *app = new ApplicationI(NULL);
        app->initialize();
        app->proxy = BaseObj::adapter->add(app, communicator->stringToIdentity("flowengine"));
        BaseObj::adapter->activate();
        communicator->waitForShutdown();
    }
    catch(const Ice::Exception& ex)
    {
        cerr << ex << endl;
        status = EXIT_FAILURE;
    }
    
```

</details>  


Server config:
```
FlowEngine.Endpoints=tcp -p 10000
```
And thats all.

**Reply by externl** [reply]
What are the endpoints of `pSession`? This proxy's endpoints are the reason for the new connection. If you're getting directly from the server adding a new servant it likely contains internal endpoints. You'll need to set the `.PublishedEndpoints` property for the server OA or only listen on the public network interface.

**Comment by Fireshtorm1k** [comment]
>endpoints that are local to the server's internal network but that the client doesn't have access to.  

Thank you for aiming.  That was problem. Server used 0.0.0.0 as host address, all resolved.
